tools/fuzzer/afl-main.c

```diff
@@ -0,0 +1,463 @@
+	if (!fp) {
+		fprintf(stderr, "error: opening ipc message file %s\n",
+			ipcmsg_filename);
+		return -errno;
+	}
+
+	/*
+	 * ipc message file contents:
+	 * header msg_size msg_data
+	 */
+
+	/* read header from file */
+	ret = fread(&(fuzzer->msg.header), sizeof(uint32_t), 1, fp);
+	if (ret != 1) {
+		fprintf(stderr, "error: reading header of ipc msg from %s\n",
+			ipcmsg_filename);
+		return -EIO;
+	}
+
+	if (afl_debug)
+		fprintf(stdout, "header read: %ud\n", fuzzer->msg.header);
+
+	/* read msg size from file */
+	ret = fread(&(fuzzer->msg.msg_size), sizeof(unsigned int), 1, fp);
+	if (ret != 1) {
+		fprintf(stderr, "error: reading msg_size of ipc msg from %s\n",
+			ipcmsg_filename);
+		return -EIO;
+	}
+
+	if (afl_debug)
+		fprintf(stdout, "msg_size read: %u\n", fuzzer->msg.msg_size);
+
+	/* read msg contents fromfile */
+	uret = fread(fuzzer->msg.msg_data, sizeof(char), fuzzer->msg.msg_size, fp);
+	if (uret != fuzzer->msg.msg_size) {
+		fprintf(stderr, "error: reading msg_data of ipc msg from %s\n",
+			ipcmsg_filename);
+		return -EIO;
+	}
+
+	fclose(fp);
+	return 0;
+}
+
+int main(int argc, char *argv[])
+{
+	struct fuzz fuzzer;
+	int ret;
+	char opt;
+	char *platform_name = NULL, *ipcmsg_file = NULL;
+	int i;
+	FILE *fp = NULL;
+
+	/* parse arguments */
+	while ((opt = getopt(argc, argv, "hi:p:")) != -1) {
+		switch (opt) {
+		case 'i':
+			ipcmsg_file = optarg;
+			break;
+		case 'p':
+			platform_name = optarg;
+			break;
+		case 'h':
+			usage(argv[0]);
+			exit(0);
+			break;
+		default:
+			usage(argv[0]);
+			exit(EXIT_FAILURE);
+		}
+	}
+
+	/* initialise emulated target device */
+	if (!platform_name) {
+		fprintf(stderr, "error: no target platform specified\n");
+		usage(argv[0]);
+		exit(EXIT_FAILURE);
+	}
+
+	/* initialise emulated target device */
+	if (!ipcmsg_file) {
+		fprintf(stderr, "error: no ipc message file specified\n");
+		usage(argv[0]);
+		exit(EXIT_FAILURE);
+	}
+
+	/* find platform */
+	for (i = 0; i < ARRAY_SIZE(platform); i++) {
+		if (!strcmp(platform[i]->name, platform_name))
+			break;
+	}
+
+	/* no platform found */
+	if (i >= ARRAY_SIZE(platform)) {
+		fprintf(stderr, "error: platform %s not supported\n",
+			platform_name);
+		usage(argv[0]);
+		exit(EXIT_FAILURE);
+	}
+
+	pid_t fid = fork();
+
+	if (!fid) {
+
+		/*
+		 * Hardcode qemu-system-xtensa path for byt platform.
+		 * TODO: path and options should be different for different platforms
+		 */
+		static char *newargv[] = {"/home/sof/work/qemu/build/xtensa-softmmu/qemu-system-xtensa",
+				    "-cpu", "baytrail", "-M", "adsp_byt", "-nographic", "-kernel",
+				    "/home/sof/work/sof.git/build_byt_gcc/src/arch/xtensa/sof-byt.ri",
+				    NULL};
+		static char *newenviron[] = { NULL };
+
+		fprintf(stdout, "Executing qemu xtensa\n");
+		execve(newargv[0], newargv, newenviron);
+
+		/* execve() returns only on error */
+		perror("execve");
+		exit(EXIT_FAILURE);
+	}
+
+	/* TODO: Need to make sure that child is executed first. */
+	sleep(5);
```

Comment:
We should probably tell the user here we are waiting N secs for qemu to start, again, I would pass the sleep tim in cmd line and make 5 secs the default.

Reply:
This is a good idea. I updated to code to take wait time as an optional argument. This can be found here. (dc53ddb71a535e6282d7a13cf78cd9b1845023f4).